api4/license.go

```diff
@@ -296,3 +301,174 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+		}
+		trueUpReviewPlugins.TotalActivePlugins = len(trueUpReviewPlugins.ActivePluginNames)
+
+		for _, plugin := range pluginResponse.Inactive {
+			trueUpReviewPlugins.InactivePluginNames = append(trueUpReviewPlugins.InactivePluginNames, plugin.Name)
+		}
+		trueUpReviewPlugins.TotalInactivePlugins = len(trueUpReviewPlugins.InactivePluginNames)
+	}
+
+	// Authentication Features
+	config := c.App.Config()
+	mfaUsed := config.ServiceSettings.EnforceMultifactorAuthentication
+	ldapUsed := config.LdapSettings.Enable
+	samlUsed := config.SamlSettings.Enable
+	openIdUsed := config.OpenIdSettings.Enable
+	guessAccessAllowed := config.GuestAccountsSettings.Enable
+
+	authFeatures := map[string]*bool{
+		model.TrueUpReviewAuthFeaturesMfa:        mfaUsed,
+		model.TrueUpReviewAuthFeaturesADLdap:     ldapUsed,
+		model.TrueUpReviewAuthFeaturesSaml:       samlUsed,
+		model.TrueUpReviewAuthFeatureOpenId:      openIdUsed,
+		model.TrueUpReviewAuthFeatureGuestAccess: guessAccessAllowed,
+	}
+
+	authFeatureList := []string{}
+	for feature, used := range authFeatures {
+		if used != nil && *used {
+			authFeatureList = append(authFeatureList, feature)
+		}
+	}
+
+	reviewProfile := model.TrueUpReviewProfile{
+		ServerId:               c.App.TelemetryId(),
+		ServerVersion:          model.CurrentVersion,
+		ServerInstallationType: os.Getenv(telemetry.EnvVarInstallType),
+		LicenseId:              license.Id,
+		LicensedSeats:          *license.Features.Users,
+		LicensePlan:            license.SkuName,
+		CustomerName:           license.Customer.Name,
+		ActiveUsers:            activeUserCount,
+		TotalIncomingWebhooks:  incomingWebhookCount,
+		TotalOutgoingWebhooks:  outgoingWebhookCount,
+		Plugins:                trueUpReviewPlugins,
+		AuthenticationFeatures: authFeatureList,
+	}
+
+	// Convert true up review profile struct to map
+	var telemetryProperties map[string]interface{}
+	reviewProfileJson, err := json.Marshal(reviewProfile)
+	if err != nil {
+		c.Err = model.NewAppError("requestTrueUpReview", "api.marshal_error", nil, "", http.StatusInternalServerError)
+		return
+	}
+
+	dueDate := utils.GetNextTrueUpReviewDueDate(time.Now())
+	status, err := c.App.Srv().Store().TrueUpReview().GetTrueUpReviewStatus(dueDate.UnixMilli())
+	if err != nil {
+		status, err = c.App.Srv().Store().TrueUpReview().CreateTrueUpReviewStatusRecord(status)
+
+		if err != nil {
+			c.Err = model.NewAppError("requestTrueUpReview", "api.license.true_up_review.create.fail.app_error", nil, "", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// Do not send true-up review data if the user has already requested one for the quarter.
+	if !status.Completed {
+		// Send telemetry data.
+		json.Unmarshal(reviewProfileJson, &telemetryProperties)
+		delete(telemetryProperties, "plugins")
+		plugins := reviewProfile.Plugins.ToMap()
+		for pluginName, pluginValue := range plugins {
+			telemetryProperties[pluginName] = pluginValue
```

Comment:
@neallred good callout!

> the plugins and auth features seem like more like telemetry data since as far as I know they aren't needed for checking license compliance

The primary objective here for us is to gather data needed to conduct quarterly true up reviews. Everything else is a bonus intended to help CSMs have more meaningful conversations with customers.

Outside of the two you mentioned (plugins and auth features), are there other things you think are under telemetry data?

Reply:
Yeah, the other ones that seem like telemetry more than strictly checking license compliance are:
* server version
* server installation type
* total incoming webhooks
* total outgoing webhooks

As mentioned above, I think even if we decide we want to collect plugin info, we'll want to do similar to how `trackPluginConfig` does it and only collect info on plugins we administer through the Mattermost marketplace.

Reply:
👍 I'll add a change to ensure we only use plugins administered through the MM Marketplace.

Reply:
In the case that we need to remove the telemetry data, I've made a PR on another branch with changes to make that happen: https://github.com/mattermost/mattermost-server/pull/21978

Removes:

- server version
- server installation type
- total incoming webhooks
- total outgoing webhooks
- plugin data
- auth features

The true up review "profile" looks like this with the change:

```json
{
	"active_users": 1,
	"customer_name": "Mattermost Dev License",
	"license_id": "m7gqopm2xs30u6q8of7bry4vel",
	"license_plan": "Enterprise Dev",
	"licensed_seats": 200000,
	"server_version": "7.7.0"
}
```

Reply:
:100:

Reply:
@ConorMacpherson @neallred 

So I heard back from legal and we're good to proceed with collecting the following information:

- Server ID
- Customer Name (if available)
- License Id
- Number of license seats bought
- License plan (e.g. Pro, Enterprise)
- License expiry date
- Number of current activated users
- Server version
- Authentication features
- Number and which Plugin/integration
- Number of webhooks

We're covered by existing legalese, specifically the Privacy Policy. Personally Identifiable Information (PII) would pertain to end user information, and not the customer at an account-level.

That said, as we will need to add some documentation on the true-up feature we've developed here, we'll list there the items we are collecting.

Let me know if there's further questions!

cc: @marianunez